File: 2024/2/Python/solution.py
import os
from typing import List, Dict

class Solution:

  def __init__(self, path):
    self._reports: List[List[int]]
    self._safe_dictionary: Dict[str, List[List[int]]]
    self._path: str = path
  
  def _is_decreasing(self, reports):
    return reports == sorted(reports, reverse=True)

  def _is_increasing(self, reports):
    return reports == sorted(reports)
  
  def _is_difference_correct(self, report):
    for idx, level in enumerate(report[:-1]):
      abs_val = abs(level - report[idx+1])
      if abs_val > 3 or abs_val == 0:
        return False
    return True

  def _ingest_input(self, path):
    with open(path, "r") as f:
      higher_list = []
      for line in f:
        lower_list = [int(letter) for letter in line.split() if letter.isdigit()]
        higher_list.append(lower_list)
      self._reports = higher_list

  def _is_safe(self):

    safe_reports = []
    unsafe_reports = []
    for report in self._reports:
      if self._is_decreasing(report) or self._is_increasing(report):
        if self._is_difference_correct(report):
          safe_reports.append(report)
        else:
          unsafe_reports.append(report)
      else:
        unsafe_reports.append(report)
    self._safe_dictionary = {"Safe":safe_reports, "Unsafe":unsafe_reports}

  def run(self):
    self._ingest_input(self._path)
    self._is_safe()
    print(self._safe_dictionary)

if __name__ == '__main__':
  input_path = os.path.dirname(os.path.realpath(__file__)).replace("Python", "input.txt")
  
  solution = Solution(input_path)
  solution.run()

File: 2024/2/Cpp/src/Solution.cpp
#include <iostream>
#include <string>
#include <fstream>
#include "../include/Solution.h"

// PRIVATE:

void Solution::ingestFile(const std::string& filePath){

    std::ifstream inputFile(filePath);

    if (!inputFile.is_open()) {
        throw std::runtime_error("Error while ingesting a file [function: ingest_file()]");
    }

    std::string line;

    while (std::getline(inputFile, line)) {
        std::vector<int> tempList;
        for (char ch : line) {
            if (std::isdigit(ch)) {
                int digit = ch - '0'; 
                // "0" represents a string object
                // while '0' represents a char in cpp
                // substracting char '0' from digit char gives us an integer (character - ASCII value of 0 = integer character)
                tempList.push_back(digit);
            }
        }
        reports.push_back(tempList);
    }

    inputFile.close();
};

bool Solution::isDecreasing(std::vector<int>& report){
    for (size_t i = 1; i < report.size(); ++i){
        if (report[i-1] < report[i]){
            return false;
        }
    }
    return true;
};

bool Solution::isIncreasing(std::vector<int>& report){
    for (size_t i = 1; i < report.size(); ++i){
        if (report[i-1] > report[i]){
            return false;
        }
    }
    return true;
};

bool Solution::isDifferenceCorrect(std::vector<int>& report){
    for (size_t i = 1; i < report.size(); ++i){
        int diff = abs(report[i-1] - report[i]);
        if (diff < 1 || diff > 3){
            return false;
        }
    }
    return true;
};

void Solution::isSafe(){
    for (size_t i = 0; i < reports.size(); ++i){
        if (isDecreasing(reports[i]) || isIncreasing(reports[i])) {
            if (isDifferenceCorrect(reports[i])){
                mappedReports["Safe"].push_back(reports[i]);
            }else{
                mappedReports["Unsafe"].push_back(reports[i]);
            }
        }else{
            mappedReports["Unsafe"].push_back(reports[i]);
        }
    }
};

// PUBLIC:

void Solution::printSafeReports(){
    std::cout << "Safe reports:" << std::endl;
    for (const auto& vec : mappedReports["Safe"]) {
        std::string tempStr;
        for (int num : vec) {
            std::string tempStr2 = std::to_string(num) + " ";
            tempStr += tempStr2;
        }
        std::cout << tempStr << std::endl;
    }

    std::cout << "Unsafe reports:" << std::endl;
    for (const auto& vec : mappedReports["Unsafe"]) {
        std::string tempStr;
        for (int num : vec) {
            std::string tempStr2 = std::to_string(num) + " ";
            tempStr += tempStr2;
        }
        std::cout << tempStr << std::endl;
    }
}

void Solution::printReports(){
    // for debugging purposes
    std::string tempStr;
    std::cout << reports.size() << std::endl;
    for (size_t i = 0; i < reports.size(); ++i){
        std::string tempStr;
        for (size_t j = 0; j < reports[i].size(); ++j) {
            std::string tempStr2 = std::to_string(reports[i][j]) + " ";
            tempStr += tempStr2;
        }
        std::cout << tempStr << std::endl;

    }
}

void Solution::run(){
    isSafe();
    printSafeReports();
    // print_reports();
};

Solution::Solution(const std::string& filePath) {
    ingestFile(filePath);
};

Solution::~Solution(){

};